src/models/eval.ts

```diff
@@ -0,0 +1,457 @@
+    }
+    return convertResultsToTable(await this.toResultsFile());
+  }
+
+  async addResult(result: EvaluateResult, columnIdx: number, rowIdx: number, test: AtomicTestCase) {
+    const newResult = await EvalResult.createFromEvaluateResult(
+      this.id,
+      result,
+      columnIdx,
+      rowIdx,
+      test,
+    );
+    this.results.push(newResult);
+  }
+
+  async addPrompts(prompts: CompletedPrompt[]) {
+    this.prompts = prompts;
+    const db = getDb();
+    await db.update(evalsTable).set({ prompts }).where(eq(evalsTable.id, this.id)).run();
+  }
+
+  async addProviders(providers: Provider[]) {
+    const db = getDb();
+    await db.transaction(async (tx) => {
+      for (const provider of providers) {
+        const id = provider.id;
+        tx.insert(evalsToProviders)
+          .values({
+            evalId: this.id,
+            providerId: id,
+          })
+          .onConflictDoNothing()
+          .run();
+      }
+    });
+  }
+
+  async loadResults() {
+    this.results = await EvalResult.findManyByEvalId(this.id);
+  }
+
+  async getResults(): Promise<EvaluateResult[] | EvalResult[]> {
+    if (this.useOldResults()) {
+      invariant(this.oldResults, 'Old results not found');
+      return this.oldResults.results;
+    }
+    await this.loadResults();
+    return this.results;
+  }
+  async toEvaluateSummary(): Promise<EvaluateSummary> {
+    if (this.useOldResults()) {
+      invariant(this.oldResults, 'Old results not found');
+      return this.oldResults;
+    }
+    if (this.results.length === 0) {
+      await this.loadResults();
+    }
+    const stats: EvaluateStats = {
+      successes: 0,
+      failures: 0,
+      tokenUsage: {
+        cached: 0,
+        completion: 0,
+        prompt: 0,
+        total: 0,
+      },
+    };
+
+    for (const prompt of this.prompts) {
+      stats.successes += prompt.metrics?.testPassCount || 0;
+      stats.failures += prompt.metrics?.testFailCount || 0;
+      stats.tokenUsage.prompt += prompt.metrics?.tokenUsage.prompt || 0;
+      stats.tokenUsage.cached += prompt.metrics?.tokenUsage.cached || 0;
+      stats.tokenUsage.completion += prompt.metrics?.tokenUsage.completion || 0;
+      stats.tokenUsage.total += prompt.metrics?.tokenUsage.total || 0;
+    }
+
+    return {
+      version: 2,
+      timestamp: new Date(this.createdAt).toISOString(),
+      results: this.results.map((r) => r.toEvaluateResult()),
+      stats,
+    };
+  }
+
+  async toResultsFile(): Promise<ResultsFile> {
+    const summary = await this.toEvaluateSummary();
+    const results: ResultsFile = {
+      version: this.version(),
+      createdAt: new Date(this.createdAt).toISOString(),
+      results: summary,
+      config: this.config,
+      author: this.author || null,
+      prompts: this.prompts,
+      datasetId: this.datasetId || null,
+    };
+
+    return results;
+  }
+
+  async delete() {
+    const db = getDb();
+    await db.transaction(() => {
```

Comment:
what do you mean?  

you will need to do it via this method because of the FK constraints, these existed before because the prompt and dataset FK's already existed

Reply:
Every time we've added a new table or foreign key we've broken delete. Comment was an instinctual reaction. This looks good.